design/secrets_provider_milestone1_2_design.md

```diff
@@ -0,0 +1,367 @@
+
+```pseudocode
+GET /secrets{?variable_ids}
+    If error
+        Add erroring secret to list
+        GET /secrets{?variable_ids} with rest of variableIDs
+        Print list of variableIDs that errored
+        return Conjur secrets
+```
+
+
+
+#### Fetching Relevant K8s Secrets
+
+With the decoupling of Secrets Provider and application pod, we can know what types of K8s Secrets the pod requires by either of the following ways: 
+
+|      | Fetch/Update K8s Secret                                      | Elaboration                                                  | Pros                                                         | Cons                                        |
+| ---- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------- |
+| 1    | Read all K8s Secrets that are marked with labels and update them with Conjur secrets |                                                              | - Less intervention on our part as a customer can define which types of secrets we should parse<br /><br />- K8s native, as-is solution | - Not generalized so we are still K8s bound |
+| 2    | Get secrets by configuration in Secrets Provider             | Define K8S_SECRETS in `cyberark-secrets-provider` manifest<br />` name: K8S_SECRET` |                                                              | - Repetitive work<br /><br />- Not scalable |
+
+To know which K8s Secrets the apps in the namespace need we believe the first solution, Read all K8s Secrets that are marked with labels, would be the most K8s native and painless solution for the customer. Also it avoids the repetitive work of having to list all the K8s Secrets each app needs.
+
+It is important to note however, that we are responsible for providing a solution for updating the K8s secrets and *not* how those Pods will get those secrets.
+
+##### **Customer Experience**
+
+K8s offers the option of adding labels on K8s secrets. For example, a customer would add a label on a K8s Secret like so:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: db-secret
+  labels:
+    key: CONJUR <----
+  type: Opaque
+stringData:
+  conjur-map: |-
+    username: secrets/my_username
+```
+
+These labels do not have to be Conjur-specific and the customer would decide which label they would like to put in the K8s secrets. They would define which secrets to iterate over by "K8S_SECRET_LABEL" env variable in their Secrets Provider manifest. The Secrets Provider will perform a search for secrets with that label against the [K8s](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#secret-v1-core)/[OC](https://docs.openshift.com/container-platform/4.4/rest_api/index.html#secret-v1-core) API and search for the "conjur_map" on the results.
```

Comment:
Why you chose Env var?
What if the customer discovered that someone else is using the same label? would it have to restart the pod?
In the future we might support rotation and this will prevent us of doing runtime changes

Reply:
> Why you chose Env var?
Do you suggest another way to retrieve this data and support runtime changes? 

What do you mean that someone else is using the same label?  IMO this isn't a different than if a new K8s secret exists and a customer would need to add it to the list of `K8S_SECRETS` in the init container solution. Regardless they will need to restart the Pod.

Reply:
what is the value of `K8S_SECRET_LABEL` in the example above? is it `conjur-secret` or `env-prod`?

Reply:
@Tovli we can move to config-map loaded as volume for live configuration.
We do not put it in this solution design because it is overwhelmed with info already.
This can be done later on.
Also, this is irrelevant for Job/init-container (which are the flavores supported by this milestone) since they are single operation anyway, so live configuration is unnecessary.